modules/core/src/test/java/org/apache/ignite/internal/processors/cache/transactions/TxSavepointsSelfTest.java

```diff
@@ -0,0 +1,305 @@
+
+                    tx.savepoint("s1");
+
+                    tx.savepoint("s1");
+                } catch (IllegalArgumentException e) {
+                    assertTrue(e.getMessage().startsWith("Savepoint \"s1\" already exist"));
+                }
+            }
+    }
+
+    /**
+     * Tests rollbacks to the same savepoint instance.
+     */
+    public void testMultipleRollbackToSavepoint() {
+        for (TransactionConcurrency concurrency : TransactionConcurrency.values())
+            for (TransactionIsolation isolation : TransactionIsolation.values()) {
+                cache.put(1, 0);
+
+                try (Transaction tx = grid(0).transactions().txStart(concurrency, isolation)) {
+                    cache.put(1, 1);
+
+                    tx.savepoint("s1");
+
+                    cache.put(1, 2);
+
+                    tx.savepoint("s2");
+
+                    cache.put(1, 3);
+
+                    tx.savepoint("s3");
+
+                    tx.rollbackToSavepoint("s2");
+
+                    assertEquals("Failed in " + concurrency + ' ' + isolation + " transaction.",
+                        (Integer) 2, cache.get(1));
+
+                    cache.put(1, 4);
+
+                    assertEquals("Failed in " + concurrency + ' ' + isolation + " transaction.",
+                        (Integer) 4, cache.get(1));
+
+                    tx.rollbackToSavepoint("s2");
+
+                    assertEquals("Failed in " + concurrency + ' ' + isolation + " transaction.",
+                        (Integer) 2, cache.get(1));
+
+                    tx.commit();
+                }
+
+                assertEquals("Failed in " + concurrency + ' ' + isolation + " transaction.",
+                    (Integer) 2, cache.get(1));
+            }
+    }
+
+    /**
+     * Tests savepoints in failed transaction.
+     */
+    public void testTransactionRollback() {
+        for (TransactionConcurrency concurrency : TransactionConcurrency.values())
+            for (TransactionIsolation isolation : TransactionIsolation.values()) {
+                try (Transaction tx = grid(0).transactions().txStart(concurrency, isolation)) {
+                    cache.put(1, 1);
+
+                    tx.savepoint("s1");
+
+                    cache.put(1, 2);
+
+                    tx.savepoint("s2");
+
+                    cache.put(1, 3);
+
+                    tx.savepoint("s3");
+
+                    tx.releaseSavepoint("s3");
+
+                    tx.rollbackToSavepoint("s2");
+
+                    assertEquals("Failed in " + concurrency + ' ' + isolation + " transaction.",
+                        (Integer) 2, cache.get(1));
+
+                    tx.rollback();
+                }
+
+                assertEquals("Failed in " + concurrency + ' ' + isolation + " transaction.",
+                    null, cache.get(1));
+            }
+    }
+
+    /**
+     * Tests two caches within one transaction.
+     */
+    public void testMultiCaches() {
```

Comment:
may be change to "testCrossCacheRollbackToSavepoint" ? Because multiple caches in one transaction is called cross cache transaction